snorkel/features/relative_features.py

```diff
@@ -0,0 +1,148 @@
+import os
+import re
+import sys
+
+sys.path.append(os.path.join(os.environ['SNORKELHOME'], 'treedlib'))
+
+from collections import defaultdict
+from entity_features import compile_entity_feature_generator, get_ddlib_feats
+from functools import partial
+from snorkel.models import Span
+from string import punctuation
+from tree_structs import corenlp_to_xmltree
+from treedlib import compile_relation_feature_generator
+from snorkel.utils import get_as_dict
+
+
+def get_span_splits(candidate, stopwords=None):
+    split_pattern = r'[\s{}]+'.format(re.escape(punctuation))
+    for i, arg in enumerate(candidate.get_contexts()):
+        for token in re.split(split_pattern, arg.get_span().lower()):
+            if stopwords is None or token not in stopwords:
+                yield 'SPAN_SPLIT[{0}][{1}]'.format(i, token), 1
+
+
+def get_span_splits_stopwords(stopwords):
+    return partial(get_span_splits, stopwords=stopwords)
+
+
+def get_span_feats(candidate, stopwords=None):
+    args = candidate.get_contexts()
+    if not isinstance(args[0], Span):
+        raise ValueError("Accepts Span-type arguments, %s-type found.")
+
+    # Unary candidates
+    if len(args) == 1:
+        get_tdl_feats = compile_entity_feature_generator()
+        span          = args[0]
+        sent          = get_as_dict(span.parent)
+        xmltree       = corenlp_to_xmltree(sent)
+        sidxs         = range(span.get_word_start(), span.get_word_end() + 1)
+        if len(sidxs) > 0:
+
+            # Add DDLIB entity features
+            for f in get_ddlib_feats(sent, sidxs):
+                yield 'DDL_' + f, 1
+
+            # Add TreeDLib entity features
+            for f in get_tdl_feats(xmltree.root, sidxs):
+                yield 'TDL_' + f, 1
+
+    # Binary candidates
+    elif len(args) == 2:
+        get_tdl_feats = compile_relation_feature_generator()
+        span1, span2  = args
+        xmltree       = corenlp_to_xmltree(get_as_dict(span1.parent))
+        s1_idxs       = range(span1.get_word_start(), span1.get_word_end() + 1)
+        s2_idxs       = range(span2.get_word_start(), span2.get_word_end() + 1)
+        if len(s1_idxs) > 0 and len(s2_idxs) > 0:
+
+            # Apply TDL features
+            for f in get_tdl_feats(xmltree.root, s1_idxs, s2_idxs,
+                                   stopwords=stopwords):
+                yield 'TDL_' + f, 1
+    else:
+        raise NotImplementedError("Only handles unary or binary candidates")
+
+
+def get_span_feats_stopwords(stopwords):
+    return partial(get_span_feats, stopwords=stopwords)
```

Comment:
This actually hurt performance on CDR. My stopword set may have been too aggressive. But we're obvs going to make that pass on TDL anyways sooooo